Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeInteractionRegionLayers.mm

```diff
@@ -192,77 +231,68 @@ void updateLayersForInteractionRegions(const RemoteLayerTreeNode& node)
-                    [regionLayer setBorderWidth:6];
-                    [regionLayer setName:@"Guard"];
-                }
-                break;
-            case InteractionRegion::Type::Interaction:
-                setInteractionRegion(regionLayer.get(), interactionRegionGroupName);
-                configureLayerForInteractionRegion(regionLayer.get(), interactionRegionGroupName);
-                if (applyBackgroundColorForDebugging) {
-                    [regionLayer setBackgroundColor:cachedCGColor({ WebCore::SRGBA<float>(0, 1, 0, .2) }).get()];
-                    [regionLayer setName:@"Interaction"];
-                }
-                break;
+        RetainPtr<CALayer> regionLayer;
+        bool didReuseLayer = true;
+        bool didReuseLayerBasedOnRect = false;
+
+        auto findOrCreateLayer = [&]() {
+            auto layerIterator = existingLayers.find(key);
+            if (layerIterator != existingLayers.end()) {
+                didReuseLayerBasedOnRect = true;
+                regionLayer = layerIterator->value;
+                return;
             }
-        }
 
-        if (!foundInPosition)
-            [layer insertSublayer:regionLayer.get() atIndex:insertionPoint];
+            auto layerReuseIterator = reusableLayers.find(reuseKey);
+            if (layerReuseIterator != reusableLayers.end()) {
+                regionLayer = layerReuseIterator->value;
+                return;
+            }
 
-        if (![interactionRegionGroupName isEqualToString:interactionRegionGroupNameForLayer(regionLayer.get())]) {
-            if (region.type == InteractionRegion::Type::Guard)
-                configureLayerAsGuard(regionLayer.get(), interactionRegionGroupName);
-            if (region.type == InteractionRegion::Type::Interaction)
-                configureLayerForInteractionRegion(regionLayer.get(), interactionRegionGroupName);
+            didReuseLayer = false;
+            if (region.type == InteractionRegion::Type::Interaction) {
+                regionLayer = adoptNS([[interactionRegionLayerClass() alloc] init]);
+                return;
+            }
+            regionLayer = adoptNS([[CALayer alloc] init]);
+        };
+        findOrCreateLayer();
+
+        if (didReuseLayer) {
+            auto layerKey = std::make_pair(enclosingIntRect([regionLayer frame]), region.type);
+            existingLayers.remove(layerKey);
+            reusableLayers.remove(reuseKey);
+
+            bool shouldReconfigureRemoteEffect = didReuseLayerBasedOnRect && ![interactionRegionGroupName isEqualToString:interactionRegionGroupNameForLayer(regionLayer.get())];
+            if (shouldReconfigureRemoteEffect)
+                configureRemoteEffect(regionLayer.get(), region.type, interactionRegionGroupName);
+        } else {
+            initializeInteractionRegionLayer(regionLayer.get(), region.type, interactionRegionGroupName);
```

Comment:
Is there a reason initializeInteractionRegionLayer isn't `createInteractionRegionLayer` and gets called instead of `regionLayer = adoptNS([[CALayer alloc] init]);`?